source/extensions/filters/listener/local_ratelimit/local_ratelimit.cc

```diff
@@ -0,0 +1,70 @@
+#include "source/extensions/filters/listener/local_ratelimit/local_ratelimit.h"
+
+#include <algorithm>
+#include <string>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/common/platform.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/network/listen_socket.h"
+#include "envoy/stats/scope.h"
+
+#include "source/common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ListenerFilters {
+namespace LocalRateLimit {
+
+Config::Config(
+    const envoy::extensions::filters::listener::local_ratelimit::v3::LocalRateLimit& proto_config,
+    Event::Dispatcher& dispatcher, Stats::Scope& scope, Runtime::Loader& runtime)
+    : enabled_(proto_config.runtime_enabled(), runtime),
+      stats_(generateStats(proto_config.stat_prefix(), scope)) {
+
+  rate_limiter_ = std::make_shared<Filters::Common::LocalRateLimit::LocalRateLimiterImpl>(
+      std::chrono::milliseconds(
+          PROTOBUF_GET_MS_REQUIRED(proto_config.token_bucket(), fill_interval)),
+      proto_config.token_bucket().max_tokens(),
+      PROTOBUF_GET_WRAPPED_OR_DEFAULT(proto_config.token_bucket(), tokens_per_fill, 1), dispatcher,
+      Protobuf::RepeatedPtrField<
+          envoy::extensions::common::ratelimit::v3::LocalRateLimitDescriptor>());
+}
+
+bool Config::canCreateConnection() { return rate_limiter_->requestAllowed({}); }
+
+Config::~Config() {
+  // Reset the rate_limiter_ first so that the weak_ptr will be expired() if this was the last
+  // reference.
+  rate_limiter_.reset();
```

Comment:
I'm unsure I understand the need to explicitly reset the shared pointer in the destructor, as it will be reset during normal clean-up.

Reply:
right we probably wont need to reset. I stole this from network local ratelimit but I thought reset is needed since the rate limiter is shared in network filter.